scripts/plot_performance_breakdown.py

```diff
@@ -1,65 +1,143 @@
 #!/usr/bin/env python3
 
-# Given a benchmark result json that was generated with --sql_metrics, this script plots the cost of the different
+# Given a benchmark result json that was generated with --metrics, this script plots the cost of the different
 # steps (parsing, optimization, ...) as a normalized stacked bar chart.
 
 import json
 import matplotlib.pyplot as plt
 import pandas as pd
-import sys
+import argparse
 
 benchmarks = []
+rule_benchmarks = []
 
-if(len(sys.argv) != 2):
-    exit("Usage: " + sys.argv[0] + " benchmark.json")
+parser = argparse.ArgumentParser(description='''Given a benchmark result json that was generated with --metrics,
+                                this script plots the cost of the different steps (parsing, optimization, ...) as a
+                                normalized stacked bar chart.''')
+parser.add_argument('filename', action='store', help='Benchmark json filename')
+parser.add_argument('--plot_rules', dest='plot_rules', action='store_true', help='Plot performance breakdown for each optimizer rule')
+options = parser.parse_args()
 
-with open(sys.argv[1]) as file:
+with open(options.filename) as file:
     data = json.load(file)
 
 for benchmark_json in data['benchmarks']:
     benchmark = []
+    rule_benchmark = []
     benchmark.append(benchmark_json['name'])
+    rule_benchmark.append(benchmark_json['name'])
 
     sum_parse_duration = 0.0
     sum_sql_translation_duration = 0.0
     sum_optimization_duration = 0.0
+    sum_optimizer_rule_durations = {}
     sum_lqp_translation_duration = 0.0
     sum_plan_execution_duration = 0.0
 
     for run in benchmark_json['successful_runs']:
         if len(run['metrics']) == 0:
-            exit("No metrics found. Did you run the benchmark with --sql_metrics?")
+            exit("No metrics found. Did you run the benchmark with --metrics?")
         for metrics in run['metrics']:
             sum_parse_duration += metrics['parse_duration']
 
             for statement in metrics['statements']:
                 sum_sql_translation_duration += statement['sql_translation_duration']
                 sum_optimization_duration += statement['optimization_duration']
+                if statement['optimizer_rule_durations']:
+                    for rule_name, rule_duration in statement['optimizer_rule_durations'].items():
+                        # crop typeid().name down to the actual rule name. This might be compiler dependent
+                        rule_name = rule_name[11:-1]
```

Comment:
Did you check gcc and clang? We could also use cxxfilt if it is an issue, but that might be overkill.

Reply:
astonishingly this works with clang-9 on mac and linux